include/ncutility/detail/BinarySerializationDetail.h

```diff
@@ -0,0 +1,483 @@
+#pragma once
+
+#include "ncutility/NcError.h"
+
+#include <algorithm>
+#include <array>
+#include <iostream>
+#include <optional>
+#include <ranges>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+/** @cond internal */
+namespace nc::serialize::binary
+{
+template<class T>
+concept TriviallyCopyable = requires { requires std::is_trivially_copyable_v<T>; };
+
+template<class T>
+concept Aggregate = requires { requires std::is_aggregate_v<T>; };
+
+// Max supported member count for automatic aggregate serialization
+inline constexpr auto g_aggregateMaxMemberCount = 16ull;
+
+// Arbitrarily large value returned from MemberCount() when a type has too many members.
+inline constexpr size_t g_failedMemberCount = 0xFFFFFFFFFFFFFFFF;
+
+// A type which is implicitly convertable to all other types
+struct UniversalType{ template<class T> operator T() const; };
+
+// Count the number of members in an aggregate, returns g_failedMemberCount for types with > 16 members
+template<class T>
+    requires requires { std::is_aggregate_v<T>; }
+consteval auto MemberCount() -> size_t
+{
+    using U = UniversalType;
+    if constexpr (requires { T{U{}, U{}, U{}, U{},  U{}, U{}, U{}, U{},  U{}, U{}, U{}, U{},  U{}, U{}, U{}, U{},  U{}}; })
```

Comment:
The spaces are a bit wonky in MemberCount

Reply:
Oh yeah, I initially made them wonky for easier counting. Don't really need to keep them that way.